lib/commands/recordscreen.js

```diff
@@ -1,61 +1,160 @@
-import _ from 'lodash';
-import { retryInterval, waitForCondition } from 'asyncbox';
-import B from 'bluebird';
-import { util, fs, tempDir } from 'appium-support';
-import { exec } from 'teen_process';
+import { fs, tempDir } from 'appium-support';
+import { SubProcess } from 'teen_process';
 import log from '../logger';
-import { getPidUsingPattern, encodeBase64OrUpload } from '../utils';
+import { encodeBase64OrUpload } from '../utils';
+import iProxy from '../wda/iproxy';
+import B from 'bluebird';
 
 
 let commands = {};
 
-const RETRY_PAUSE = 1000;
 const MAX_RECORDING_TIME_SEC = 60 * 10;
 const DEFAULT_RECORDING_TIME_SEC = 60 * 3;
-const PROCESS_SHUTDOWN_TIMEOUT_SEC = 5;
-const REAL_DEVICE_BINARY = 'xrecord';
-const REAL_DEVICE_PGREP_PATTERN = (udid) => `${REAL_DEVICE_BINARY}.*${udid}`;
-const SIMULATOR_BINARY = 'xcrun';
-const SIMULATOR_PGREP_PATTERN = (udid) => `simctl io ${udid} recordVideo`;
-const DEFAULT_EXT = '.mp4';
-
-async function extractCurrentRecordingPath (pid) {
-  const {output} = await exec('ps', ['o', 'command', '-p', pid]);
-  log.debug(`Got the following output from ps: ${output}`);
-  const pattern = new RegExp(/[\s="'](\/.*\.mp4)/);
-  const matches = pattern.exec(output);
-  return _.isEmpty(matches) ? null : _.last(matches);
-}
+const DEFAULT_MJPEG_SERVER_PORT = 9100;
+const MP4_EXT = '.mp4';
+const DEFAULT_FPS = 10;
 
-async function finishScreenCapture (pid) {
-  try {
-    await exec('kill', ['-2', pid]);
-  } catch (e) {
-    return true;
+
+class ScreenRecorder {
+  constructor (udid, videoPath, opts = {}) {
+    this.videoPath = videoPath;
+    this.opts = opts;
+    this.udid = udid;
+    this.binaryName = 'ffmpeg';
+    this.mainProcess = null;
+    this.iproxy = null;
+    this.timeoutHandler = null;
   }
-  try {
-    await waitForCondition(async () => {
+
+  async start (timeout) {
+    try {
+      await fs.which(this.binaryName);
+    } catch (err) {
+      throw new Error(`'${this.binaryName}' binary is not found in PATH. Install it using 'brew install ffmpeg'` +
+        `Check https://www.ffmpeg.org/download.html for more details.`);
+    }
+
+    const localPort = this.opts.remotePort;
+    if (this.opts.usePortForwarding) {
+      await this.startIproxy(localPort);
+    }
+
+    const args = [
+      '-f', 'mjpeg',
+      '-r', this.opts.videoFps,
+      '-i', `http://localhost:${localPort}`,
+      '-vcodec', 'mjpeg',
+      '-y', this.videoPath,
+    ];
+    this.mainProcess = new SubProcess(this.binaryName, args);
+    this.mainProcess.on('output', (stdout, stderr) => {
+      if (stderr && !stderr.includes('frame=')) {
+        log.warn(`[${this.binaryName}] ${stderr}`);
+      }
+    });
+    await this.mainProcess.start(0);
+    // Give ffmpeg some time for init
+    await B.delay(5000);
```

Comment:
Why not just use the process start wait? https://github.com/appium/node-teen_process/blob/master/lib/subprocess.js#L37

Reply:
because I usually don't know for which exact line to wait

Reply:
But `await  this.mainProcess.start(5000);` is essentially the same as what you are doing.

Reply:
ah, will change